Task: BJS2-41377

Darkselect opened the pull request:
Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 1.0

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 1.0

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 2.0

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 3.0

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 4.0

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 5.0

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 6.0

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 7.0

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 8.0

DeemaP commented on src/main/resources/application.yaml:
```diff
@@ -22,8 +22,10 @@ spring:
       port: 6379
       host: localhost
       channel:
+        ad_bought: ad_bought_channel
```
Вот о чем я в пост-сервисе тебе написал. Там используется вкладка name, тут нет. Давай одинаково писать)

Darkselect replied:
исправил

DeemaP commented on src/main/java/faang/school/analytics/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Bean
+    public RedisTemplate<String, Object> redisTemplate(RedisConnectionFactory connectionFactory) {
```
Пока мы отсюда ничего не отправляем, нам темплейт-то и не нужен.
Но даже если и оставлять его здесь, подскажи, насколько корректно, что в redisContainer ты суешь JedisConnectionFactory, а в template ты кладешь RedisConnectionFactory.

Darkselect replied:
Переделал

DeemaP commented on src/main/java/faang/school/analytics/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,51 @@
+        RedisTemplate<String, Object> redisTemplate = new RedisTemplate<>();
+        redisTemplate.setConnectionFactory(connectionFactory);
+        redisTemplate.setKeySerializer(new StringRedisSerializer());
+        redisTemplate.setValueSerializer(new GenericToStringSerializer<>(AdBoughtEventListener.class));
```
Подскажи почему в serializer ты пробрасываешь AdBoughtEventListener.class?

Darkselect replied:
Переделал

DeemaP commented on src/main/java/faang/school/analytics/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,51 @@
+        return redisTemplate;
+    }
+
+
```
лишний отступ

DeemaP commented on src/main/java/faang/school/analytics/listener/AdBoughtEventListener.java:
```diff
@@ -0,0 +1,42 @@
+        this.analyticsEventService = analyticsEventService;
+        this.analyticsEventMapper = analyticsEventMapper;
+    }
+
```
лишний отступ

DeemaP commented on src/main/java/faang/school/analytics/model/AnalyticsEvent.java:
```diff
@@ -33,4 +34,10 @@ public class AnalyticsEvent {
     @Column(name = "received_at", nullable = false)
     private LocalDateTime receivedAt;
+
+    @Column(name = "payment_amount")
```
Насколько вообще корректно добавлять сюда эти колонки? У нас в аналитике только сохранялось кто с кем и что сделал. Нужно ли расширять таблицу настолько, что у нас в будет таблица неимоверных размеров и все будет лежать где-попало.

Нельзя же все в одном месте хранить. Иначе мы сюда скоро все банковские счета и переводы привяжем и ваще будет адский ад.

Darkselect replied:
Это требовалось по условие задачи. Есть предложения сделать иначе ?

Darkselect replied:
В событии содержатся: id поста; id пользователя; сумма платежа; длительность рекламы; дата и время.

DeemaP replied:
Не посмотрел условие, сори, да. Ок, значит пусть будет так. Хотя конечно как по мне, в аналитике должна лежать только простая инфа
(Кто, с кем, что и когда сделал), а если тебе подробности нужны, идти и вытаскивать из соответствующих сервисов нужные данные. Но условие задачи есть условие задачи)

А уж если хочется в аналитике подробные данные по всему собирать, то надо делать доп репозитории под каждый тип событий, которые нам будут нужны с подробной аналитикой по ним)

Но условие есть условие, ок)

DeemaP commented on src/main/java/faang/school/analytics/mapper/AnalyticsEventMapper.java:
```diff
@@ -1,12 +1,27 @@
+                .receiverId(event.getActorId())
+                .paymentAmount(event.getPaymentAmount())
+                .adDuration(event.getAdDuration())
+                .receivedAt(LocalDateTime.now())
```
Давайте инфу о дате какого-то события передавать из сервиса, откуда у нас пришло событие?

Сейчас получается, что у тебя там кто-то купил подписку, ты в базу сохранил время покупки в том сервисе, потом передал ивент, он летел сколько-то времени (возможно завис), прилетел сюда, а здесь на него совсем другой таймстамп встает. Нелогично получается. Можно же прям в json передать таймстаймп корректный и везде будет одно время.

Darkselect replied:
переделал

DeemaP commented on src/main/java/faang/school/analytics/listener/AdBoughtEventListener.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Override
+    protected void saveEvent(AdBoughtEventResponseDto event) {
```
Давай метод называть handleEvent (мы же не только сохраняем, а сначала его обрабатываем)

Darkselect replied:
Ок

DeemaP reviewed (commented):
Тут побольше вопросиков осталось)
Мне кажется перемудрил слегка с расширением сущности.

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 9.0

drewbal commented on src/main/java/faang/school/analytics/deserializer/BigDecimalDeserializer.java:
```diff
@@ -0,0 +1,20 @@
+public class BigDecimalDeserializer extends JsonDeserializer<BigDecimal> {
+    @Override
+    public BigDecimal deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
+        JsonNode node = p.getCodec().readTree(p);
```
расскажешь как это работает? продвинутый уровень

Darkselect replied:
Мы используем этот BigDecimalDeserializer для десериализации JSON-значений в объекты BigDecimal. Это необходимо, когда мы получаем JSON-данные из внешнего источника и хотим преобразовать их в Java-объекты.

drewbal commented on src/main/java/faang/school/analytics/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Bean
+    public RedisTemplate<String, Object> redisTemplate() {
```
а это нужно? ты как будто его нигде не используешь

drewbal commented on src/main/java/faang/school/analytics/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,49 @@
+        RedisTemplate<String, Object> redisTemplate = new RedisTemplate<>();
+        redisTemplate.setConnectionFactory(jedisConnectionFactory());
+        redisTemplate.setKeySerializer(new StringRedisSerializer());
+        redisTemplate.setValueSerializer(new GenericToStringSerializer<>(Object.class));
```
зачем в String, если в другом сервисе ты делаешь JSON? В чем логика? Поясни хитрость

drewbal commented on src/main/java/faang/school/analytics/deserializer/BigDecimalDeserializer.java:
```diff
@@ -0,0 +1,20 @@
+    @Override
+    public BigDecimal deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
+        JsonNode node = p.getCodec().readTree(p);
+        if (node.isArray() && node.size() == 2) {
```
что это за magic number?

Darkselect replied:
Добавил константы

drewbal commented on src/main/java/faang/school/analytics/deserializer/BigDecimalDeserializer.java:
```diff
@@ -0,0 +1,20 @@
+
+public class BigDecimalDeserializer extends JsonDeserializer<BigDecimal> {
+    @Override
+    public BigDecimal deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
```
а где исключение обрабатывается? Не просто так же ты его выбрасываешь? Не хватает ума сообразить задумку

Darkselect replied:
Сделал через try catch и добавил логи

drewbal commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -0,0 +1,46 @@
+        try {
+            return objectMapper.readValue(message.getBody(), eventType);
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to map message to event type", e);
```
а логов не нужно?

Darkselect replied:
Забыыыыл :)

drewbal commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -0,0 +1,46 @@
+
+@Slf4j
+@RequiredArgsConstructor
+public abstract class AbstractListener<T> implements MessageListener {
```
Надо, чтобы ты рассказал про это чудо

Darkselect replied:
Так Влад же рассказывал про это. В своих лекциях. Мы все можем абстрагировать вот так

drewbal commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -0,0 +1,46 @@
+    protected final ObjectMapper objectMapper;
+    private final List<EventHandler<T>> eventHandlers;
+
+    protected T listenEvent(Message message, Class<T> eventType) throws IOException {
```
а почему protected?

Darkselect replied:
Да чисто мой бзик, потому что будем же наследоваться от этого класса.

drewbal commented on src/main/resources/application.yaml:
```diff
@@ -22,8 +22,10 @@ spring:
       port: 6379
       host: localhost
       channel:
+        ad_bought: ad_bought_channel
```
kebab-case

Darkselect replied:
исправил

drewbal commented on src/main/java/faang/school/analytics/dto/analyticsEvent/AdBoughtEventResponseDto.java:
```diff
@@ -0,0 +1,26 @@
+@Data
+@Builder
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AdBoughtEventResponseDto {
```
почему бы ее не назвать так же, как ты делал в post_service? А то как будто разные люди писали

Darkselect replied:
Дима назвал свою дто-ху так. Я подумал, что именно так и нужно называть тут дто-ху. Но согласен да переименую.

Darkselect replied:
Я думал тут какой то стандарт

DeemaP replied:
Ты погляди где используется AnalyticsEventResponseDto, она же даже называется response, то есть я ее возвращаю в запросе, у нас в аналитик сервисе есть всего один эндпоинт, в котором можно получить всю аналитику по receiverId и типу Ивента.

Например в эндпоинт кинуть айди юзера 1 и тип FOLLOW, в результате мы получаем лист AnalyticsEventResponseDto, в котором лежат все события по подписках на нашего юзера 1 (кто и когда это сделал)

drewbal commented on src/main/java/faang/school/analytics/mapper/AnalyticsEventMapper.java:
```diff
@@ -1,12 +1,27 @@
 
+import java.time.LocalDateTime;
+
 @Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
```
объясни в чем смысл маппинга? ты получил событие event через редис, в нем есть вся информация. Зачем ее маппить в аналогичный файл?

Darkselect replied:
Если я уберу это, то результаты будет непредсказуемыми

drewbal commented on src/test/java/faang/school/analytics/listener/AdBoughtEventListenerTest.java:
```diff
@@ -0,0 +1,84 @@
+    private AdBoughtEventListener listener;
+
+    @Test
+    public void testListenEvent() throws IOException {
```
а негативные исходы ты не тестируешь? правильно ли исключения вылетают и метод отрабатывает, если что-то идет не так?

drewbal reviewed (commented):
Я бы сказал гениально, но далеко не все понял. Весьма высокий уровень, я запутался в нем. Но некоторые решения кажутся странными

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 11.0

RuslanZinovyev commented on src/main/java/faang/school/analytics/deserializer/BigDecimalDeserializer.java:
```diff
@@ -0,0 +1,33 @@
+import java.math.BigDecimal;
+
+@Slf4j
+public class BigDecimalDeserializer extends JsonDeserializer<BigDecimal> {
```
Нам действительно нужен кастомный десериалайзер тут?
Можем мы сделать что-то наподобие
```
ObjectMapper mapper = new ObjectMapper();
mapper.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);
```

RuslanZinovyev commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -0,0 +1,59 @@
+        if (message.getBody().length == 0) {
+            log.error("Message body is empty {}", message.getBody());
+            throw new IOException("Message body is empty");
+        } else {
```
для чего тут нужен else, какую роль он выполняет?

RuslanZinovyev commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -0,0 +1,59 @@
+        try {
+            T event = listenEvent(message, eventType());
+
+            if (eventHandlers != null && eventHandlers.isEmpty()) {
```
Здесь логическая ошибка, условие `eventHandlers.isEmpty()` должно быть инвертировано `!eventHandlers.isEmpty()`

RuslanZinovyev commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -0,0 +1,59 @@
+            return objectMapper.readValue(message.getBody(), eventType);
+        } catch (IOException e) {
+            log.error("Failed to map message to event type", e);
+            throw new RuntimeException("Failed to map message to event type", e);
```
```
log.error("Failed to map message to event type", e);
        throw new EventDeserializationException("Failed to map message to event type", e);
```
создай кастомное исключение

```
public class EventDeserializationException extends RuntimeException {
    public EventDeserializationException(String message, Throwable cause) {
        super(message, cause);
    }
}
```

RuslanZinovyev commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -0,0 +1,59 @@
+
+        } catch (IOException e) {
+            log.error("Failed to process event", e);
+            throw new EventProcessingException("Failed to process event");
```
`throw new EventProcessingException("Failed to process event", e);`

RuslanZinovyev commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    protected abstract Class<T> eventType();
```
здесь можно попробовать убрать abstract и добавить реализацию через рефлексию

```
protected Class<T> eventType() {
    ParameterizedType type = (ParameterizedType) getClass().getGenericSuperclass();
    return (Class<T>) type.getActualTypeArguments()[0];
}
```

в таком случае все дочерние классы могут использовать метод родительского класса напрямую без обеспечения реализации, при условии что они будут реализованы следующим образом

`public class MyEventListener extends AbstractListener<MyEvent> { }`

RuslanZinovyev commented on src/main/java/faang/school/analytics/listener/AdBoughtEventListener.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    protected void handleEvent(AdBoughtEvent event) {
+        AnalyticsEvent analyticsEvent = analyticsEventMapper.dtoToEntity(event);
```
нужно логирование тут добавить, ты вешаешь аннотацию для логера, но фактически не используешь ее

`log.info("Processing AdBoughtEvent with postId: {}", event.getPostId());`

RuslanZinovyev commented on src/main/java/faang/school/analytics/listener/EventHandler.java:
```diff
@@ -0,0 +1,5 @@
+package faang.school.analytics.listener;
+
+public interface EventHandler <T> {
```
лишний отступ

Darkselect replied:
Нету тут лишнего отступа

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 12.0

Darkselect committed:
Merge branch 'Dragon-master-stream7' of https://github.com/CorporationX/analytics_service into Dragon-stream7-BJS2-41377

# Conflicts:
#	src/main/java/faang/school/analytics/AnalyticsServiceApp.java
#	src/main/java/faang/school/analytics/mapper/AnalyticsEventMapper.java
#	src/main/java/faang/school/analytics/service/AnalyticsEventService.java
#	src/main/resources/application.yaml
#	src/test/java/faang/school/analytics/service/AnalyticsEventServiceTest.java

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 13.0

RuslanZinovyev commented on src/main/java/faang/school/analytics/config/redis/RedisConfigProperties.java:
```diff
@@ -3,11 +3,7 @@
-        Channel channel
-) {
-    public record Channel(String adBought, String profileView) {
+public record RedisConfigProperties(String host, int port, Channel channel) {
```
Я не против records, но нам нужно выбрать одну стратегию настройки конфигурации и придерживаться ее.

RuslanZinovyev commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -20,19 +21,20 @@ protected T listenEvent(Message message, Class<T> eventType) throws IOException
         if (message.getBody().length == 0) {
             log.error("Message body is empty {}", message.getBody());
             throw new IOException("Message body is empty");
```
```
if (message.getBody().length == 0) {
    log.error("Message body is empty: {}", message.getBody());
    throw new InvalidMessageException("Message body is empty");
}
```

RuslanZinovyev commented on src/main/java/faang/school/analytics/listener/AbstractListener.java:
```diff
@@ -20,19 +21,20 @@ protected T listenEvent(Message message, Class<T> eventType) throws IOException
     }
 
-    protected abstract Class<T> eventType();
+    protected Class<T> eventType() {
```
Тщательно проверь работу этого метода как и всего класса

Darkselect replied:
Так работает же, в бд сохраняется все четко

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 14.0

Darkselect committed:
Task: BJS2-41377

Description:
Implemented a Redis-based Pub/Sub mechanism to send AdBoughtEvent from post_service to analytics_service. In post_service, the AdBoughtEventPublisher publishes events with details such as post ID, user ID, payment amount, ad duration, and timestamp when a user purchases an ad. In analytics_service, the AdBoughtEventListener listens for these events, maps them to AnalyticsEvent using AnalyticsEventMapper, and saves the data in the database through AnalyticsEventService.

Version 15.0
